File: python/next_permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        if len(nums) <= 1:
            return

        i = len(nums) -1
        j = i - 1
        while nums[i] <= nums[j] and j >=0:
            j -= 1
            i -= 1

        if j >= 0:
            i = len(nums) -1
            while nums[j] >= nums[i]:  # find a smallest num on right but bigger than nums[j] (merely bigger than nums[j])
                i -= 1
            nums[i], nums[j] = nums[j], nums[i]
            k = j +1
        else:
            k = 0  # if j ==-1, reverse all
        
        i = len(nums) -1
        while k < i:
            nums[k], nums[i] = nums[i], nums[k]
            k += 1
            i -= 1

File: cpp/next_permutation.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        // Find peek.
        int p;
        for (p = nums.size()-1; p > 0; p--)
            if (nums[p] > nums[p-1]) break;
        if (p == 0) {	// restart
            std::reverse(nums.begin(), nums.end());
            return;
        }
        
        // Find a num from peek right that mearly larger than peek's left.
        int j = p - 1;
        for (int i = nums.size()-1; i > j; i--)
            if (nums[i] > nums[j]) { // swap
            	swap(nums[i], nums[j]);
                break;
            }

        // Reverse
        std::reverse (nums.begin() + p, nums.end());
    }
};

void printVec(const std::vector<int>& v) {
	for (int val: v)
		cout << val << " ";
	cout << endl;
}

int main(int argc, char const *argv[])
{
	vector<int> nums = {1, 2, 3, 4};
	Solution s;
	for (int i = 0; i < 24; i++) {
		s.nextPermutation(nums);
		printVec(nums);
	}
	return 0;
}